packages/backend/src/apps/toolbox/common/get-for-each-variables.ts

```diff
@@ -0,0 +1,132 @@
+import { FOR_EACH_INPUT_SOURCE, FOR_EACH_ITERATION_KEY } from './constants'
+
+type InputSource = FOR_EACH_INPUT_SOURCE | null
+
+interface ProcessedInput {
+  iterations: number
+  processedItems: any
+  inputSource: InputSource
+}
+
+interface ProcessedColumn {
+  id: string
+  name: string
+  value: string
+}
+
+export interface MultipleRowObject {
+  rows: {
+    data: Record<string, string | number | null>
+    rowId?: string
+  }[]
+  columns: {
+    id: string
+    name: string
+    value: string
+  }[]
+}
+
+const ULID_REGEX = /^[0-9A-HJKMNP-TV-Z]{26}$/i
+const UUID_REGEX =
+  /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/
+
+export function isCheckboxItems(items: string[]): boolean {
+  return Array.isArray(items) && items.every((item) => typeof item === 'string')
+}
+
+function processColumns(data: MultipleRowObject): {
+  inputSource: InputSource
+  processedColumns: ProcessedColumn[]
+} {
+  if (data.columns.length === 0) {
+    return {
+      inputSource: null,
+      processedColumns: [],
+    }
+  }
+
+  // NOTE: we can tell if its a tile column by its column id
+  // tiles will either have uuid or ulid as column id
+  const hasTileColumn = data.columns.every(
+    (column: any) => ULID_REGEX.test(column.id) || UUID_REGEX.test(column.id),
+  )
+  const inputSource = hasTileColumn
+    ? FOR_EACH_INPUT_SOURCE.TILES
+    : FOR_EACH_INPUT_SOURCE.M365_EXCEL
+
+  const processedColumns = data.columns.map((column: any) => ({
+    id: column.id,
+    name: column.name,
+    value: `items.rows.${FOR_EACH_ITERATION_KEY}.data.${column.id}`,
+  }))
+
+  // NOTE: only tiles will have rowId
+  if (inputSource === FOR_EACH_INPUT_SOURCE.TILES) {
+    processedColumns.push({
+      id: 'rowId',
+      name: 'Row ID',
+      value: `items.rows.${FOR_EACH_ITERATION_KEY}.rowId`,
+    })
+  }
+
+  return { inputSource, processedColumns }
+}
+
+export function processItems(items: MultipleRowObject): ProcessedInput {
+  const { inputSource, processedColumns } = processColumns(items)
+  const processedItems = {
+    rows: items.rows,
+    columns: processedColumns,
+  }
+
+  return {
+    iterations: items.rows.length,
+    processedItems,
+    inputSource,
+  }
+}
```

Comment:
this works but seems a bit fragile, since it solely depends on the shape of the data. why not return the "inputSource" directly from the action since find multiple rows are essentially new actions. 

The other way is to read the appKey/eventKey directly from the stepId in the variable {{step.xxxx...}}.


For checkbox input, i think there's no other way, but maybe call it `stringArray` or sth instead of checkbox so it's more extensible.